python/mxnet/image.py

```diff
@@ -437,19 +441,53 @@ def next(self):
         try:
             while i < batch_size:
                 label, s = self.next_sample()
-                data = [imdecode(s)]
-                if len(data[0].shape) == 0:
-                    logging.debug('Invalid image, skipping.')
+                data = [self.imdecode(s)]
+                try:
+                    self.check_valid_image(data)
+                except RuntimeError, e:
+                    logging.debug('Invalid image, skipping:  %s' % str(e))
                     continue
-                for aug in self.auglist:
-                    data = [ret for src in data for ret in aug(src)]
-                for d in data:
+                data = self.prepare_augmentation(data)
+                for datum in data:
                     assert i < batch_size, 'Batch size must be multiples of augmenter output length'
-                    batch_data[i][:] = nd.transpose(d, axes=(2, 0, 1))
+                    batch_data[i][:] = self.preprocess_data(datum)
                     batch_label[i][:] = label
                     i += 1
         except StopIteration:
             if not i:
                 raise StopIteration
 
         return io.DataBatch([batch_data], [batch_label], batch_size-i)
+    
+    
+    def check_data_shape(self, data_shape):
+        """Checks that the input data shape is valid"""
+        if not len(data_shape) == 3:
+            raise ValueError('data_shape should have length 3, with dimensions CxHxW')
+        if not data_shape[0] == 3:
+            raise ValueError('This iterator expects inputs to have 3 channels.')
+
+    def check_valid_image(self, data):
+        """Checks that data is valid"""
+        if len(data[0].shape) == 0:
+            raise RuntimeError('Data shape is wrong')
+        
+    def imdecode(self, s):
+        """Decodes a sting or byte string into an image."""
+        return imdecode(s)
+    
+    def read_image(self, fname):
+        """Reads image from fname and returns the raw bytes to be decoded."""
+        with open(os.path.join(self.path_root, fname), 'rb') as fin:
+            img = fin.read()
+        return image
+    
+    def prepare_augmentation(self, data):
```

Comment:
its actually doing the augumentation, not preparing for it. How about transform.